### PR TITLE
vSphere problem detector

### PR DESCRIPTION
**Pull request by jsafrane**
Here's proposal to check if OpenShift clusters installed into vSphere environment is / can be fully functional by introducing a new operator and periodical checks, so our customers / our support can quickly check for common issues in OCP / vSphere config.

Goal: reduce nr. BZs. Out of 69 bugs that mentioned "vSphere" or "VMware", 32 were closed as some mis-configuration / wrong documentation.

WIP:
 * finalize naming
 * add some tests to the installer for pre-flight checks.
 * see Open Questions

### CODE REVIEW

File: enhancements/storage/vsphere-storage-check.md
Comment (jcpowermac) [inline enhancements/storage/vsphere-storage-check.md:40]: @gnufied with your recent converstation with vmware folks I thought vmotion wasn't an issue. We know that **storage** vmotion is not supported

Comment (gnufied) [reply enhancements/storage/vsphere-storage-check.md:40]: Yep, it might be trickier to detect something like that in a monitoring tool though...

File: enhancements/storage/vsphere-storage-check.md
Comment (jcpowermac) [inline enhancements/storage/vsphere-storage-check.md:81]: This should be possible with events

File: enhancements/storage/vsphere-storage-check.md
Comment (jcpowermac) [inline enhancements/storage/vsphere-storage-check.md:83]: This as well should be fairly simple to grab off the config from the cluster. (if the user has access)

File: enhancements/storage/vsphere-storage-check.md
Comment (jcpowermac) [inline enhancements/storage/vsphere-storage-check.md:94]: Add ESXi host version because this could be a mixture.

File: enhancements/storage/vsphere-storage-check.md
Comment (jcpowermac) [reply enhancements/storage/vsphere-storage-check.md:40]: If storage drs is enable we could know that. If a storage vmotion we would just have tasks and events to monitor.

File: enhancements/storage/vsphere-storage-check.md
Comment (jsafrane) [reply enhancements/storage/vsphere-storage-check.md:97]: Added

File: enhancements/storage/vsphere-storage-check.md
Comment (JoelSpeed) [inline enhancements/storage/vsphere-storage-check.md:42]: I've been looking at some conversations around our CI env and as far as I could tell, the AWS VMC product that we are using for the CI clusters, enables DRS.

Where does this "not supported" come from and is this going to cause an issue for the CI?

File: enhancements/storage/vsphere-storage-check.md
Comment (staebler) [inline enhancements/storage/vsphere-storage-check.md:80]: These both seem like reasonable things to check in the installer in the "Platform Permissions Check" asset.

File: enhancements/storage/vsphere-storage-check.md
Comment (staebler) [inline enhancements/storage/vsphere-storage-check.md:83]: If there is a limit to the size of the datastore name that OCP can handle, then the installer should be validating the length as part of validating the install config. For the StorageClasses and PVs, however, that does not seem like something that the installer could check as they do not exist until the cluster is created.

File: enhancements/storage/vsphere-storage-check.md
Comment (gnufied) [inline enhancements/storage/vsphere-storage-check.md:131]: should we call it `VMWareStorageMonitoring`? I would prefer if we kept it as `vSphereMonitoring` may be.. because at the end of the day we are monitoring more than just storage..

**Review by staebler — COMMENTED** [review]
The proposed pre-flight checks for the installer seem reasonable to me.

**Review by JoelSpeed — COMMENTED** [review]
I'm gonna start by saying this is a fantastic idea and I think will help many teams more than just storage!

Which leads me on to my next point, this is currently very much focused on the storage aspects of vSphere and the misconfigurations that have caused storage team bugs.

I wonder if this could become slightly more generic (eg the CRD naming which currently contains storage) so that it wouldn't feel weird if, for example, the Machine API team wanted to add some checks using the same operator to check the prerequisites for our components to work. I feel like this could definitely become a shared component that reports back on aspects of the environment for multiple teams to digest.

File: enhancements/storage/vsphere-storage-check.md
Comment (gnufied) [inline enhancements/storage/vsphere-storage-check.md:153]: Also lets use product name as a prefix rather than the vendor..

File: enhancements/storage/vsphere-storage-check.md
Comment (jcpowermac) [reply enhancements/storage/vsphere-storage-check.md:42]: @JoelSpeed yeah DRS is enabled, not sure we could even disable it in VMC.
Since early September
Total vMotion Migrations: 7055

Comment (jcpowermac) [reply enhancements/storage/vsphere-storage-check.md:42]: **Storage** DRS is _not_ enabled.

Comment (JoelSpeed) [reply enhancements/storage/vsphere-storage-check.md:42]: This highlights my lack of awareness of vSphere, I didn't realise there were two different types of DRS!
So if I understand, there is DRS for VMs and DRS for Storage? We have the VM one working and that's not causing issues? But Storage DRS isn't on in VMC?

Comment (jcpowermac) [reply enhancements/storage/vsphere-storage-check.md:42]: > This highlights my lack of awareness of vSphere, I didn't realise there were two different types of DRS!
> So if I understand, there is DRS for VMs and DRS for Storage? 

Right

> We have the VM one working and that's not causing issues? 

Don't get me wrong there are still openshift-test failures in VMC. We should add more montoring to determine if vmotion would be a potential cause of a test failure. I tend to think there are more issues with the vSphere cluster is busy.

I did create this script to determine if there was a correlation between test failures and vmotion 
https://gist.github.com/jcpowermac/9246a8030afe8af15cc2e913d88ca3fb
need to do more testing with it to have a better answer.

> But Storage DRS isn't on in VMC?

Right, I haven't found good documentation for vSAN with SDRS but from comments it doesn't seem to be supported from VMware.

File: enhancements/storage/vsphere-storage-check.md
Comment (elmiko) [inline enhancements/storage/vsphere-storage-check.md:36]: is there a missing part to this sentence?

> that is either mis-configured.

seems like there should be an "or ..."

File: enhancements/storage/vsphere-storage-check.md
Comment (elmiko) [inline enhancements/storage/vsphere-storage-check.md:45]: minor nit

s/loose/lose/

File: enhancements/storage/vsphere-storage-check.md
Comment (elmiko) [inline enhancements/storage/vsphere-storage-check.md:50]: imo, it would also be nice to know about their topology and features in use.

**Review by elmiko — COMMENTED** [review]
i think this is an interesting idea, i agree with @JoelSpeed (and others) who are suggesting this operator could expand to cover more than just storage. i added a couple suggestions inline.

File: enhancements/storage/vsphere-storage-check.md
Comment (jsafrane) [reply enhancements/storage/vsphere-storage-check.md:43]: For a VM migration, I don't see an event generated at all. And for storage migration, our VMC does not allow that (there's only one datastore)

File: enhancements/storage/vsphere-storage-check.md
Comment (jsafrane) [reply enhancements/storage/vsphere-storage-check.md]: OCP docs do not show "list events" as necessary privilege: https://docs.openshift.com/container-platform/4.6/installing/installing_vsphere/installing-vsphere-installer-provisioned.html#installation-vsphere-installer-infra-requirements_installing-vsphere-installer-provisioned

Should we add it?

File: enhancements/storage/vsphere-storage-check.md
Comment (jsafrane) [reply enhancements/storage/vsphere-storage-check.md:131]: Sure, changed to `VSphereMonitoring`.

File: enhancements/storage/vsphere-storage-check.md
Comment (jsafrane) [reply enhancements/storage/vsphere-storage-check.md:83]: Split into two checks, Datastore in cloud config (checked in the installer) and StorageClasses/PVs.

File: enhancements/storage/vsphere-storage-check.md
Comment (jsafrane) [reply enhancements/storage/vsphere-storage-check.md:153]: Fixed

File: enhancements/storage/vsphere-storage-check.md
Comment (jsafrane) [reply enhancements/storage/vsphere-storage-check.md:36]: fixed

File: enhancements/storage/vsphere-storage-check.md
Comment (jsafrane) [reply enhancements/storage/vsphere-storage-check.md:45]: fixed

File: enhancements/storage/vsphere-storage-check.md
Comment (jsafrane) [reply enhancements/storage/vsphere-storage-check.md:50]: What features are we interested in? From storage POV, we know that only storage DRS is problematic.

**Comment by jsafrane** [comment]
> this is currently very much focused on the storage aspects of vSphere and the misconfigurations that have caused storage team bugs.

@JoelSpeed, I am happy to add more checks for non-storage related issues. Do you have any?

In addition, the enhancement expects that the monitoring operator runs in an storage namespace + the ehancement is in storage dir, should we move it somewhere else? Where? This may have consequence on changes in some other operator that should start the vSphere monitoring when the right cloud is detected. Storage team is still interested in writing the operator itself (or at least contributing heavily).

**Comment by jsafrane** [comment]
/hold
to remember to squash the commits before merge

File: enhancements/storage/vsphere-storage-check.md
Comment (Danil-Grigorev) [reply enhancements/storage/vsphere-storage-check.md:43]: @jsafrane There is one upstream metric which was merged yesterday, which will allow tracking control-plane node (only) migrations between vCenters in a multi-vcenter cluster configuration. Just did a cherry pick on that - https://github.com/openshift/kubernetes/pull/447 But the main goal was to show vCenter version, this is just a side-effect... Is this something that this proposal might need?

**Review by michaelgugino — CHANGES_REQUESTED** [review]
I don't have any issues with the suggested operator, but we are putting the cart before the horse if all of this stuff is not clearly documented.  Whatever conditions we're checking for, the docs need to land first.

**Comment by jcpowermac** [comment]
I know this is quite a bit ...

Are these features enabled?
- HA
- DRS
- SDRS w/DatastoreCluster

Types:
- Storage (NFS, vSAN, FC, iSCSI)
- Virtual Switch (Standard, Distributed, NSX)

Networking:
- If using a standard vSwitch port group is each host configured the same?

Cluster state:
- vSphere license type
- vCenter version, _each_ host version
- vCenter topology from cluster to top-level parent.
- Are there hosts in maintenance mode?
- Is NTP (or PTP) enabled, running and configured. Does each host have time that is generally in sync?
- Are hosts overcommitted?

RHCOS virtual machines:
- Tasks and events
- CPU ready, disk latency

Storage:
- Virtual machines per datastore
- FC{,oE} or iSCSI pathing issues, dead luns
- NFS?
- vSAN?
- Lost access to volume?

**Comment by jcpowermac** [comment]
Another check for the installer since I just ran into and I have seen other people have issues as well:
```
[11-11-20 16:06:06] Uploading O11N_VA-8.2.0.12949-16995073-system.vmdk... Error: Post "https://10.3.32.8/nfc/52193b60-245f-c716-c68e-ade9a7ce3af6/disk-0.vmdk": dial tcp 10.3.32.8:443: i/o timeout
govc: Post "https://10.3.32.8/nfc/52193b60-245f-c716-c68e-ade9a7ce3af6/disk-0.vmdk": dial tcp 10.3.32.8:443: i/o timeout
```
Do I have connectivity to the ESXi hosts to upload the OVA?

**Comment by jcpowermac** [comment]
I might have missed it but it seems a permissions check should be in here too. That probably should be done in the installer but I can see that also checked continously.

**Comment by jsafrane** [comment]
@michaelgugino, we already document permissions that OCP needs: https://docs.openshift.com/container-platform/4.6/installing/installing_vsphere/installing-vsphere-installer-provisioned.html#installation-vsphere-installer-infra-requirements_installing-vsphere-installer-provisioned and we're fixing it in https://github.com/openshift/openshift-docs/pull/27328

**Comment by jsafrane** [comment]
I added more metrics from @jcpowermac list, however, I'd like to start small before the metrics explode.

As the time for 4.7 runs out quickly, I propose:
* The operator lives is `openshift-cluster-storage-operator` namespace, fully managed by cluster-storage-operator. In the future, we may move it somewhere else - nothing in the cluster depends on it.
* The name is `vsphere-problem-detector`. It may be hard to rename it to say `cloud-problem-detector` later! It will get recorded as github repo name, in various places in our build pipeline and image names!

File: enhancements/storage/vsphere-problem-detector.md
Comment (jcpowermac) [inline enhancements/storage/vsphere-problem-detector.md:99]: We should be able to get vibs, assuming the user would have rights to query that.

```javascript
hostLength = clusterComputeResource.host.length

for (i =0; i < hostLength; i++) {
    System.log(clusterComputeResource.host[i].summary.config.product.version)
    System.log(clusterComputeResource.host[i].summary.config.product.build)

    softwarePackages = clusterComputeResource.host[i].configManager.imageConfigManager.fetchSoftwarePackages()

    System.log(softwarePackages.length)

    for(j=0;j<softwarePackages.length;j++){
        System.log(softwarePackages[j].name)
    }
}
```

**Comment by JoelSpeed** [comment]
> * The name is `vsphere-problem-detector`. It may be hard to rename it to say `cloud-problem-detector` later! It will get recorded as github repo name, in various places in our build pipeline and image names!

I think this is fine, if we decide we want to extend in the future to support other clouds we can create a new project and then integrate the vSphere stuff into it, though at present, seems to be mostly focused on vSphere. Would rather we didn't block on getting something like this started because we are fussing over potential future scenarios!

File: enhancements/storage/vsphere-storage-check.md
Comment (jsafrane) [reply enhancements/storage/vsphere-storage-check.md]: Yes, vCenter version should be reported - I added it to the list of metrics.

**Comment by jsafrane** [comment]
Squashed. Let the final review round begin.
/hold cancel

File: enhancements/storage/vsphere-problem-detector.md
Comment (JoelSpeed) [inline enhancements/storage/vsphere-problem-detector.md:9]: Should these not be GitHub usernames? I was under the impression that was the convention at least

**Review by JoelSpeed — COMMENTED** [review]
LGTM, just had one question about the metadata but can be fixed at some other point

File: enhancements/storage/vsphere-problem-detector.md
Comment (jsafrane) [reply enhancements/storage/vsphere-problem-detector.md:9]: They probably should...

File: enhancements/storage/vsphere-problem-detector.md
Comment (smarterclayton) [inline enhancements/storage/vsphere-problem-detector.md:132]: Why is it standalone?  What operator installs it?

Why is this not part of the vsphere machine api operator?

File: enhancements/storage/vsphere-problem-detector.md
Comment (smarterclayton) [inline enhancements/storage/vsphere-problem-detector.md:134]: Why is this a function of the storage operator?  How many vsphere problems have to do with storage vs api vs lb vs machines?

File: enhancements/storage/vsphere-problem-detector.md
Comment (jsafrane) [reply enhancements/storage/vsphere-problem-detector.md:132]: One line below it says it's installed by CSO

File: enhancements/storage/vsphere-problem-detector.md
Comment (jsafrane) [reply enhancements/storage/vsphere-problem-detector.md:134]: Because so far we have only storage problems to check, see above. It would detect ~40%-50% of storage problems we have on vSphere, where users can make lot of mistakes.

Comment (jsafrane) [reply enhancements/storage/vsphere-problem-detector.md:134]: BTW, we can move it somewhere else if it's useful.

**Comment by sttts** [comment]
Was this self-approved (by automation) ? Was it intentional?

**Comment by elmiko** [comment]
i _think_ PRs here are auto-approved. i've opened  a PR on this repo and had it approved immediately but i never investigated the mechanisms behind it.

**Comment by jsafrane** [comment]
@sttts I got agreement from Cloud API & installer team, perhaps not as /approve, still IMO valid.

**Pull request merged**
